src/core/ext/filters/http/message_compress/compression_filter.cc

```diff
@@ -0,0 +1,312 @@
+  if (max_recv_message_length.has_value() &&
+      message->payload()->Length() >
+          static_cast<size_t>(*max_recv_message_length)) {
+    return absl::ResourceExhaustedError(absl::StrFormat(
+        "Received message larger than max (%u vs. %d)",
+        message->payload()->Length(), *max_recv_message_length));
+  }
+  // Check if decompression is enabled (if not, we can just pass the message
+  // up).
+  if (!enable_decompression_ ||
+      (message->flags() & GRPC_WRITE_INTERNAL_COMPRESS) == 0) {
+    return std::move(message);
+  }
+  // Try to decompress the payload.
+  SliceBuffer decompressed_slices;
+  if (grpc_msg_decompress(algorithm, message->payload()->c_slice_buffer(),
+                          decompressed_slices.c_slice_buffer()) == 0) {
+    return absl::InternalError(
+        absl::StrCat("Unexpected error decompressing data for algorithm ",
+                     CompressionAlgorithmAsString(algorithm)));
+  }
+  // Swap the decompressed slices into the message.
+  message->payload()->Swap(&decompressed_slices);
+  message->mutable_flags() &= ~GRPC_WRITE_INTERNAL_COMPRESS;
+  return std::move(message);
+}
+
+class CompressionFilter::DecompressLoop {
+ public:
+  explicit DecompressLoop(CompressionFilter* filter, CallArgs& call_args)
+      : filter_(filter),
+        mapper_(PipeMapper<MessageHandle>::Intercept(
+            *call_args.incoming_messages)) {}
+
+  // Once we have a compression algorithm we can construct the decompression
+  // loop.
+  auto TakeAndRun(grpc_compression_algorithm algorithm) {
+    // Configure max receive size.
+    auto max_recv_message_length = filter_->max_recv_size_;
+    const MessageSizeParsedConfig* limits =
+        MessageSizeParsedConfig::GetFromCallContext(
+            GetContext<grpc_call_context_element>(),
+            filter_->message_size_service_config_parser_index_);
+    if (limits != nullptr && limits->max_recv_size().has_value() &&
+        (!max_recv_message_length.has_value() ||
+         *limits->max_recv_size() < *max_recv_message_length)) {
+      max_recv_message_length = *limits->max_recv_size();
+    }
+    // Interject decompression into the message loop.
+    return mapper_.TakeAndRun([algorithm, max_recv_message_length,
+                               filter = filter_](MessageHandle message) {
+      return filter->DecompressMessage(std::move(message), algorithm,
+                                       max_recv_message_length);
+    });
+  }
+
+ private:
+  CompressionFilter* filter_;
+  PipeMapper<MessageHandle> mapper_;
+};
+
+class CompressionFilter::CompressLoop {
+ public:
+  explicit CompressLoop(CompressionFilter* filter, CallArgs& call_args)
+      : filter_(filter),
+        mapper_(PipeMapper<MessageHandle>::Intercept(
+            *call_args.outgoing_messages)) {}
+
+  // Once we're ready to send initial metadata we can construct the compression
+  // loop.
+  auto TakeAndRun(grpc_metadata_batch& outgoing_metadata) {
```

Comment:
Same thing here about a comment indicating the return type.

Reply:
done